components/vr/webvr_thread.rs

```diff
@@ -0,0 +1,349 @@
+    contexts: HashSet<PipelineId>,
+    constellation_chan: Sender<ConstellationMsg>,
+    vr_compositor_chan: WebVRCompositorSender,
+    polling_events: bool,
+    presenting: HashMap<u64, PipelineId>
+}
+
+impl WebVRThread {
+    fn new(receiver: IpcReceiver<WebVRMsg>,
+           sender: IpcSender<WebVRMsg>,
+           constellation_chan: Sender<ConstellationMsg>,
+           vr_compositor_chan: WebVRCompositorSender)
+           -> WebVRThread {
+        let mut service = VRServiceManager::new();
+        service.register_defaults();
+        WebVRThread {
+            receiver: receiver,
+            sender: sender,
+            service: service,
+            contexts: HashSet::new(),
+            constellation_chan: constellation_chan,
+            vr_compositor_chan: vr_compositor_chan,
+            polling_events: false,
+            presenting: HashMap::new()
+        }
+    }
+
+    pub fn spawn(constellation_chan: Sender<ConstellationMsg>,
+                 vr_compositor_chan: WebVRCompositorSender)
+                 -> IpcSender<WebVRMsg> {
+        let (sender, receiver) = ipc::channel().unwrap();
+        let sender_clone = sender.clone();
+        thread::Builder::new().name("WebVRThread".into()).spawn(move || {
+            WebVRThread::new(receiver, sender_clone, constellation_chan, vr_compositor_chan).start();
+        }).expect("Thread spawning failed");
+        sender
+    }
+
+    fn start(&mut self) {
+        while let Ok(msg) = self.receiver.recv() {
+            match msg {
+                WebVRMsg::RegisterContext(context) => {
+                    self.handle_register_context(context);
+                    self.schedule_poll_events();
+                },
+                WebVRMsg::UnregisterContext(context) => {
+                    self.handle_unregister_context(context);
+                },
+                WebVRMsg::PollEvents(sender) => {
+                    self.poll_events(sender);
+                },
+                WebVRMsg::GetVRDisplays(sender) => {
+                    self.handle_get_displays(sender);
+                    self.schedule_poll_events();
+                },
+                WebVRMsg::GetFrameData(pipeline_id, device_id, near, far, sender) => {
+                    self.handle_framedata(pipeline_id, device_id, near, far, sender);
+                },
+                WebVRMsg::ResetPose(pipeline_id, device_id, sender) => {
+                    self.handle_reset_pose(pipeline_id, device_id, sender);
+                },
+                WebVRMsg::RequestPresent(pipeline_id, device_id, sender) => {
+                    self.handle_request_present(pipeline_id, device_id, sender);
+                },
+                WebVRMsg::ExitPresent(pipeline_id, device_id, sender) => {
+                    self.handle_exit_present(pipeline_id, device_id, sender);
+                },
+                WebVRMsg::CreateCompositor(device_id) => {
+                    self.handle_create_compositor(device_id);
```

Comment:
when does the compositor get disengaged if not needed for browsing context (non-VR) anymore?